File: impl/cpp/mrv_mt.cpp
// mrv parallelized over puzzles: worker threads pull (rep, puzzle) tasks
// off a shared atomic counter (dynamic scheduling; puzzle cost varies).
#include <array>
#include <atomic>
#include <bit>
#include <charconv>
#include <chrono>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <span>
#include <string>
#include <string_view>
#include <thread>
#include <vector>

namespace {

using Grid = std::array<std::uint8_t, 81>;

constexpr std::uint16_t ALL = 0x3FE;  // bits 1..9 set

constexpr auto make_table(auto f)
{
    std::array<std::uint8_t, 81> t{};
    for (int i = 0; i < 81; i++)
        t[i] = static_cast<std::uint8_t>(f(i));
    return t;
}
constexpr auto ROW = make_table([](int i) { return i / 9; });
constexpr auto COL = make_table([](int i) { return i % 9; });
constexpr auto BOX = make_table([](int i) { return i / 27 * 3 + i % 9 / 3; });

struct Solver {
    std::array<std::uint16_t, 9> rows{}, cols{}, boxes{};
    std::array<std::uint8_t, 81> empties{};
    std::uint8_t* g = nullptr;
    int n = 0;

    bool bt(int k)
    {
        if (k == n)
            return true;
        int best_j = k, best_n = 10;
        std::uint16_t best_cand = 0;
        for (int j = k; j < n; j++) {
            const int i = empties[j];
            const auto cand = static_cast<std::uint16_t>(
                rows[ROW[i]] & cols[COL[i]] & boxes[BOX[i]]);
            const int nc = std::popcount(cand);
            if (nc < best_n) {
                if (nc == 0)
                    return false;  // dead end
                best_j = j;
                best_n = nc;
                best_cand = cand;
                if (nc == 1)
                    break;
            }
        }
        const int i = empties[best_j];
        empties[best_j] = empties[k];
        empties[k] = static_cast<std::uint8_t>(i);
        const int r = ROW[i], c = COL[i], b = BOX[i];
        std::uint16_t cand = best_cand;
        while (cand) {
            const auto bit = static_cast<std::uint16_t>(cand & -cand);
            cand ^= bit;
            g[i] = static_cast<std::uint8_t>(std::countr_zero(bit));
            rows[r] ^= bit; cols[c] ^= bit; boxes[b] ^= bit;
            if (bt(k + 1))
                return true;
            rows[r] ^= bit; cols[c] ^= bit; boxes[b] ^= bit;
        }
        g[i] = 0;
        return false;
    }
};

bool solve(Grid& g)
{
    Solver s;
    s.rows.fill(ALL);
    s.cols.fill(ALL);
    s.boxes.fill(ALL);
    s.g = g.data();
    for (int i = 0; i < 81; i++) {
        if (const int d = g[i]) {
            const auto bit = static_cast<std::uint16_t>(1u << d);
            const int r = ROW[i], c = COL[i], b = BOX[i];
            if (!(s.rows[r] & s.cols[c] & s.boxes[b] & bit))
                return false;  // duplicate clue
            s.rows[r] ^= bit; s.cols[c] ^= bit; s.boxes[b] ^= bit;
        } else {
            s.empties[s.n++] = static_cast<std::uint8_t>(i);
        }
    }
    return s.bt(0);
}

long parse_arg(std::string_view s)
{
    long v = 0;
    std::from_chars(s.data(), s.data() + s.size(), v);
    return v < 1 ? 1 : v;
}

std::vector<Grid> read_puzzles()
{
    std::string input;
    char buf[1 << 16];
    std::size_t got;
    while ((got = std::fread(buf, 1, sizeof buf, stdin)) > 0)
        input.append(buf, got);

    std::vector<Grid> puzzles;
    std::string_view rest{input};
    while (!rest.empty()) {
        const auto nl = rest.find('\n');
        std::string_view line = rest.substr(0, nl);
        rest.remove_prefix(nl == std::string_view::npos ? rest.size() : nl + 1);
        if (line.ends_with('\r'))
            line.remove_suffix(1);
        if (line.empty() || line.front() == '#' || line.size() != 81)
            continue;
        Grid g;
        bool ok = true;
        for (int i = 0; i < 81; i++) {
            ok &= line[i] >= '0' && line[i] <= '9';
            g[i] = static_cast<std::uint8_t>(line[i] - '0');
        }
        if (ok)
            puzzles.push_back(g);
    }
    return puzzles;
}

void write_output(std::span<const Grid> sol, std::span<const std::uint8_t> ok)
{
    std::string out;
    out.reserve(sol.size() * 82);
    for (std::size_t p = 0; p < sol.size(); p++) {
        if (ok[p]) {
            for (const std::uint8_t d : sol[p])
                out.push_back(static_cast<char>('0' + d));
            out.push_back('\n');
        } else {
            out += "UNSOLVED\n";
        }
    }
    std::fwrite(out.data(), 1, out.size(), stdout);
}

struct Tasks {
    std::span<const Grid> puzzles;
    std::span<Grid> sol;
    std::span<std::uint8_t> ok;
    std::atomic<std::size_t> next{0};
    std::size_t total = 0;    // reps * npuz
    std::size_t lastrep = 0;  // first task index of the final repetition
};

// Only the final repetition is recorded, so every result slot is written
// exactly once and the workers never race on the same element.
void worker(Tasks& t)
{
    const std::size_t npuz = t.puzzles.size();
    for (;;) {
        const std::size_t task =
            t.next.fetch_add(1, std::memory_order_relaxed);
        if (task >= t.total)
            return;
        const std::size_t p = task % npuz;
        Grid g = t.puzzles[p];  // fresh copy each repetition
        const bool ok = solve(g);
        if (task >= t.lastrep) {
            t.ok[p] = static_cast<std::uint8_t>(ok);
            t.sol[p] = g;
        }
    }
}

}  // namespace

int main(int argc, char** argv)
{
    const long reps = argc > 1 ? parse_arg(argv[1]) : 1;
    const long threads = argc > 2 ? parse_arg(argv[2]) : 1;

    const std::vector<Grid> puzzles = read_puzzles();
    const std::size_t n = puzzles.size();
    std::vector<Grid> sol(n);
    std::vector<std::uint8_t> ok(n, 0);

    const std::size_t total = static_cast<std::size_t>(reps) * n;
    Tasks tasks{.puzzles = puzzles,
                .sol = sol,
                .ok = ok,
                .total = total,
                .lastrep = total - n};

    const auto t0 = std::chrono::steady_clock::now();
    {
        // main thread works too; threads are created once for all reps
        std::vector<std::jthread> pool;
        pool.reserve(static_cast<std::size_t>(threads - 1));
        for (long i = 1; i < threads; i++)
            pool.emplace_back(worker, std::ref(tasks));
        worker(tasks);
    }  // jthread destructors join here
    const auto t1 = std::chrono::steady_clock::now();

    write_output(sol, ok);
    const auto ns =
        std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();
    std::fprintf(stderr, "ns=%lld\n", static_cast<long long>(ns));
    return 0;
}

File: impl/go/mrv_mt.go
// mrv parallelized over puzzles: worker goroutines pull indices off an
// atomic counter (dynamic scheduling; puzzle difficulty is very uneven).
package main

import (
	"bufio"
	"fmt"
	"io"
	"math/bits"
	"os"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

const all = 0x3FE // bits 1..9 set

var boxOf [81]uint8

func init() {
	for i := 0; i < 81; i++ {
		boxOf[i] = uint8(i/27*3 + i%9/3)
	}
}

type solver struct {
	g       *[81]uint8
	rows    [9]uint16
	cols    [9]uint16
	boxes   [9]uint16
	empties [81]uint8
	n       int
}

func (s *solver) solve(g *[81]uint8) bool {
	s.g = g
	for i := 0; i < 9; i++ {
		s.rows[i], s.cols[i], s.boxes[i] = all, all, all
	}
	s.n = 0
	for i := 0; i < 81; i++ {
		d := g[i]
		if d != 0 {
			bit := uint16(1) << d
			r, c, b := i/9, i%9, boxOf[i]
			if s.rows[r]&s.cols[c]&s.boxes[b]&bit == 0 {
				return false // duplicate clue
			}
			s.rows[r] ^= bit
			s.cols[c] ^= bit
			s.boxes[b] ^= bit
		} else {
			s.empties[s.n] = uint8(i)
			s.n++
		}
	}
	return s.bt(0)
}

func (s *solver) bt(k int) bool {
	if k == s.n {
		return true
	}
	bestJ := k
	bestN := 10
	var bestCand uint16
	for j := k; j < s.n; j++ {
		i := int(s.empties[j])
		cand := s.rows[i/9] & s.cols[i%9] & s.boxes[boxOf[i]]
		nc := bits.OnesCount16(cand)
		if nc < bestN {
			if nc == 0 {
				return false // dead end
			}
			bestJ, bestN, bestCand = j, nc, cand
			if nc == 1 {
				break
			}
		}
	}
	i := int(s.empties[bestJ])
	s.empties[bestJ] = s.empties[k]
	s.empties[k] = uint8(i)
	r, c, b := i/9, i%9, boxOf[i]
	cand := bestCand
	for cand != 0 {
		bit := cand & -cand
		cand -= bit
		s.g[i] = uint8(bits.TrailingZeros16(bit))
		s.rows[r] ^= bit
		s.cols[c] ^= bit
		s.boxes[b] ^= bit
		if s.bt(k + 1) {
			return true
		}
		s.rows[r] ^= bit
		s.cols[c] ^= bit
		s.boxes[b] ^= bit
	}
	s.g[i] = 0
	return false
}

func readPuzzles() [][81]uint8 {
	data, _ := io.ReadAll(os.Stdin)
	var grids [][81]uint8
	for _, ln := range strings.Split(string(data), "\n") {
		ln = strings.TrimSpace(ln)
		if ln == "" || ln[0] == '#' || len(ln) != 81 {
			continue
		}
		var g [81]uint8
		ok := true
		for i := 0; i < 81; i++ {
			c := ln[i]
			if c < '0' || c > '9' {
				ok = false
				break
			}
			g[i] = c - '0'
		}
		if ok {
			grids = append(grids, g)
		}
	}
	return grids
}

func writeResults(results [][81]uint8, solved []bool) {
	w := bufio.NewWriter(os.Stdout)
	var line [82]byte
	line[81] = '\n'
	for i := range results {
		if solved[i] {
			for j := 0; j < 81; j++ {
				line[j] = '0' + results[i][j]
			}
			w.Write(line[:])
		} else {
			w.WriteString("UNSOLVED\n")
		}
	}
	w.Flush()
}

func main() {
	reps := 1
	threads := 1
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil && v > 0 {
			reps = v
		}
	}
	if len(os.Args) > 2 {
		if v, err := strconv.Atoi(os.Args[2]); err == nil && v > 0 {
			threads = v
		}
	}
	runtime.GOMAXPROCS(threads)
	grids := readPuzzles()
	n := len(grids)
	results := make([][81]uint8, n)
	solved := make([]bool, n)
	t0 := time.Now()
	for rep := 0; rep < reps; rep++ {
		if threads <= 1 {
			var s solver
			for i := range grids {
				g := grids[i] // fresh copy per repetition
				solved[i] = s.solve(&g)
				results[i] = g
			}
		} else {
			var next int64
			var wg sync.WaitGroup
			for w := 0; w < threads; w++ {
				wg.Add(1)
				go func() {
					defer wg.Done()
					var s solver
					for {
						idx := atomic.AddInt64(&next, 1) - 1
						if idx >= int64(n) {
							return
						}
						g := grids[idx] // fresh copy per repetition
						solved[idx] = s.solve(&g)
						results[idx] = g
					}
				}()
			}
			wg.Wait()
		}
	}
	ns := time.Since(t0).Nanoseconds()
	fmt.Fprintf(os.Stderr, "ns=%d\n", ns)
	writeResults(results, solved)
}
